### PR TITLE
Train rework

### LINKED ISSUE #76: training step refactor and testing
The actual training step should be factored into its own function and tested to check that gradient accumulation is working as expected.

### LINKED ISSUE #78: isolate and test actual training parts of train_step in their own function
train_step should be a function that just performs the actual training step - other logic should be moved elsewhere/renamed. This function should be aggressively tested for correctness.

### LINKED ISSUE #79: test and validate gradient accumulation logic (+ fix if necessary)
Check that gradient accumulation logic works as expected. The code as is looks kind of suspect at the moment.

### LINKED ISSUE #81: Checkpoint logic refactoring
Code and documentation should reflect that evaluation 'follows' checkpoints (e.g. eval is a part of the checkpoint step, not vice versa)

Checkpointing should be separated from train_step (see #78)

### LINKED ISSUE #85: simplify batching logic
Rework training batches to just use a per-epoch-generated index instead of a generator, simplify related logic

### LINKED ISSUE #86: support arbitrary tokenized dataset
are we hard-coding our tiny stories tokenized dataset atm?

### PR DESCRIPTION
**Pull request by jaidhyani**
Addressing assorted issues in [train_rework](https://github.com/delphi-suite/delphi/labels/train_rework)

### CODE REVIEW

**Comment by jaidhyani** [comment]
@jettjaniak I think it might make sense to merge this as-is, or even to break it into smaller PRs. It's already pretty big, and some of the upcoming changes are likely to be even-bulkier-by-line-count-if-not-complexity.

File: src/delphi/train/checkpoint_step.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def run_checkpoint(
+    config: GigaConfig,
+    mts: ModelTrainingState,
+    iteration_params: IterationParams,
+    train_ds: Dataset,
+    validation_ds: Dataset,
+    run_context: RunContext,
+):
+    model = mts.model
```
Comment (jettjaniak) [inline src/delphi/train/checkpoint_step.py:33]: docstring

File: src/delphi/train/config/data_config.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from beartype import beartype
+
+from delphi import constants
+
+
+@beartype
+@dataclass(frozen=True)
+class DataConfig:
+    train_dataset: str = field(
+        default=constants.TOKENIZED_CORPUS_DATASET,
```
Comment (jettjaniak) [inline src/delphi/train/config/data_config.py]: either don't set or change the name of the constant to stories-something

File: src/delphi/train/config/gigaconfig.py
Comment (jettjaniak) [inline src/delphi/train/config/gigaconfig.py:1]: change the name of gigaconfig

File: src/delphi/train/train_step.py [hunk 3]
```diff
@@ -1,110 +1,69 @@
...
     model_training_state: ModelTrainingState,
     train_ds: Dataset,
-    validation_ds: Dataset,
-    iteration_params: IterationParams,
-    eval_callbacks: list[Callable],
     config: GigaConfig,
-    train_batch_iter: Generator,
-    run_context: RunContext,
+    device: torch.device,
+    indices: list[int],
```
Comment (jettjaniak) [inline src/delphi/train/train_step.py]:
```suggestion
    ds_indices: list[int],
```

File: src/delphi/train/train_step.py [hunk 4]
```diff
@@ -1,110 +1,69 @@
...
 ):
     """
     Runs a training step, updating (mutating in place) model_training_state
```
Comment (jettjaniak) [inline src/delphi/train/train_step.py]: explain what training step entails (here, or above the relevant code blocks)

File: src/delphi/train/utils.py [hunk 5]
```diff
@@ -171,48 +172,47 @@ def initialize_model_training_state(
...
+    """ """
+    num_indices = dataset_size // batch_size
+    indices = list(range(num_indices))
+    shuffle_list(indices, seed=ordering_seed + epoch)
+    return indices
 
 
-def get_next_xy(
-    train_batch_iter: Generator, device: torch.device
+def get_xy_batch(
```
Comment (jettjaniak) [inline src/delphi/train/utils.py:198]: ds[start:end] bad, be more random

File: src/delphi/train/train_step.py [hunk 6]
```diff
@@ -1,110 +1,69 @@
...
-        for micro_step in range(config.optimizer.gradient_accumulation_steps):
-            X, Y = get_next_xy(train_batch_iter, run_context.device)
-            loss = (
-                model(X, labels=Y, return_dict=True).loss
-                / config.optimizer.gradient_accumulation_steps
+        batches = (
+            get_xy_batch(
+                dataset=train_ds,
+                indices=indices,
+                batch_size=config.batch_size,
```
Comment (jettjaniak) [inline src/delphi/train/train_step.py]: effective batch size in the config (just call it batch_size?), divisible by grad acc steps

File: src/delphi/train/training.py [hunk 7]
```diff
@@ -46,57 +63,73 @@ def run_training(config: GigaConfig) -> tuple[ModelTrainingState, RunContext]:
...
+        split=config.data_config.train_split,
+        tokens_feature=config.data_config.train_feature,
+        limit=config.data_config.train_sample_limit,
     )
-    validation_ds = cast(
-        Dataset,
-        load_delphi_training_dataset("validation", limit=config.val_sample_limit),
+    validation_ds = load_tokens_dataset_from_huggingface(
+        dataset=(
+            config.data_config.validation_dataset or config.data_config.train_dataset
```
Comment (jettjaniak) [inline src/delphi/train/training.py:71]: more explicit with variable names?
delphi-suites/dataset-name - we can call it dataset_name (HF is doing that) or dataset_id

Comment (jettjaniak) [reply src/delphi/train/training.py:71]: maybe add a comment as well

File: src/delphi/train/training.py [hunk 8]
```diff
@@ -46,57 +63,73 @@ def run_training(config: GigaConfig) -> tuple[ModelTrainingState, RunContext]:
...
+            dataset_size=len(train_ds),
+            batch_size=config.batch_size,
+            epoch=epoch,
+            ordering_seed=config.batch_ordering_seed,
         )
         model_training_state.epoch = epoch
         for step in tqdm(range(iteration_params.num_steps)):
             model_training_state.step = step
+            if should_run_checkpoint(config, model_training_state):
+                run_checkpoint(
```
Comment (jettjaniak) [inline src/delphi/train/training.py]: more like save_checkpoint?

Comment (jettjaniak) [reply src/delphi/train/training.py]: or log_and_save

File: src/delphi/train/utils.py [hunk 9]
```diff
@@ -288,3 +301,26 @@ def save_results(
...
+) -> Dataset:
+    """Load a dataset from huggingface"""
+    ds = cast(
+        Dataset,
+        load_dataset(
+            dataset,
+            split=split,
+            features=datasets.Features(
+                {tokens_feature: datasets.Sequence(datasets.Value("int32"))}
+            ),
```
Comment (jettjaniak) [inline src/delphi/train/utils.py:356]: comment, simplify?

File: tests/train/test_train_step.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
+# TODO: there are some ugly hacks here, and the test states are way too complicated
+# clean this up as other parts of the codebase are refactored
+
+from dataclasses import asdict
+
+import pytest
+import torch
+from dacite import from_dict
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]:
```suggestion
import dacite
```

File: tests/train/test_train_step.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    load_tokens_dataset_from_huggingface,
+)
+
+
+def test_accumulate_gradients_accumulates():
+    """
+    check that gradient accumulation works as expected and doesn't reset on each microstep
+    """
+    # setup
+    model = load_preset("debug").model_config.get_model()
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]: I think we were discussing removing get_model(), flagging in case that got lost in the multitude of comments

File: tests/train/test_train_step.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    dataset = load_tokens_dataset_from_huggingface(
+        dataset=constants.TOKENIZED_CORPUS_DATASET,
+        split="validation",
+        tokens_feature="tokens",
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]: this could be just "feature"

File: tests/train/test_train_step.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+    grad_a = sum(
+        [param.grad.norm() for param in model.parameters() if param.grad is not None]
+    )
+
+    # reset grad on model
+    for param in model.parameters():
+        if param.grad is not None:
+            param.grad.zero_()
+            param.grad = None
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]:
```suggestion
    model.zero_grad()
```

File: tests/train/test_train_step.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+            device=torch.device("cpu"),
+        )
+        for microstep in range(len(indices_set_b))
+    ]
+
+    # accumulate
+    total_loss = accumulate_gradients(model, batches_a, len(batches_a))
+
+    grad_a = sum(
+        [param.grad.norm() for param in model.parameters() if param.grad is not None]
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]:
```suggestion
        param.grad.norm() for param in model.parameters() if param.grad is not None
```

Comment (jettjaniak) [reply tests/train/test_train_step.py]: alternatively you could .clone() and then do torch.isclose() on the pairs

File: tests/train/test_train_step.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+    )
+
+    # reset grad on model
+    for param in model.parameters():
+        if param.grad is not None:
+            param.grad.zero_()
+            param.grad = None
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]: model.zero_grad()

File: tests/train/test_train_step.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def test_accumulate_gradients_accumulates():
+    """
+    check that gradient accumulation works as expected and doesn't reset on each microstep
+    """
+    # setup
+    model = load_preset("debug").model_config.get_model()
+    dataset = load_tokens_dataset_from_huggingface(
+        dataset=constants.TOKENIZED_CORPUS_DATASET,
```
Comment (jettjaniak) [inline tests/train/test_train_step.py]: I'm thinking we shouldn't be downloading a HF dataset for these tests. If someone wants to clone the repo, run the tests and train models on a different dataset, then it'll clone the whole train & validation stories dataset for no reason.

Comment (jettjaniak) [reply tests/train/test_train_step.py]: we just need some random ints wrapped in Dataset class, I think tests for token map are doing this

File: tests/train/test_train_step.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        best_val_loss=float("inf"),
+        last_training_step_time=0.0,
+    )
+
+
+def test_train_step_no_training():
+    """
+    Test train_step when no_training is set to True
+    """
+    # setup
```
Comment (jettjaniak) [inline tests/train/test_train_step.py:213]: fixture?

File: tests/train/test_train_step.py
Comment (jettjaniak) [inline tests/train/test_train_step.py:1]: we probably want some small reproducibility test, like compare if model initialized with a given seed after a single training step has norm of parameters close to specific value

Comment (jettjaniak) [reply tests/train/test_train_step.py:1]: not sure norm will be precise enough to fail on regressions

**Review by jettjaniak — APPROVED** [review]
get_next_xy batch should be changed, not necessarily in this PR

File: tests/train/test_train_step.py (hunk 11 above)
Comment (jaidhyani) [reply tests/train/test_train_step.py]: it's in the next PR I think

File: tests/train/test_train_step.py (hunk 16 above)
Comment (jaidhyani) [reply tests/train/test_train_step.py]: Thanks - I wanted to do something simpler here, but I was rushing and didn't want to spend the time to work out how to build a Dataset object from scratch. I'll use this.

File: tests/train/test_train_step.py
Comment (jaidhyani) [reply tests/train/test_train_step.py:1]: We could pick out a handful of specific values to test. Alternatively, and somewhat less legibly, we should store the full expected state in a static file and use it for the test comparison.
